### PR TITLE
encryption: tidy encryption slices and supply basements for KMS on Azure

### PR DESCRIPTION
**Pull request by LykxSassinator**
### What is changed and how it works?

Issue Number: ref #14693 

What's Changed:

```commit-message
This pr includes:
* Tidy the previous implementation on KMS part.
* Add credentials to make preparations for supporting encryption at rest on Azure.
```

### Related changes

- PR to update `pingcap/docs`/`pingcap/docs-cn`:
- Need to cherry-pick to the release branch

### Check List 

Tests 

- Unit test
- Integration test
- Manual test (add detailed scripts or steps below)
- No code

Side effects

- Performance regression
    - Consumes more CPU
    - Consumes more MEM
- Breaking backward compatibility

### Release note 

```release-note
None.
```

### CODE REVIEW

**Comment by wuhuizuo** [comment]
/review try to find the major problems.

**Comment by wuhuizuo** [comment]
/review review it and find the potential problems, also give some fix suggestions.

**Comment by wuhuizuo** [comment]
/review review it and find the potential problems, also give some fix suggestions.

File: components/cloud/azure/src/token_credentials_with_cache/certificate_credentials.rs [hunk 1]
```diff
@@ -0,0 +1,283 @@
...
+        "\"{}\"",
+        base64::encode(cert.to_pem().map_err(openssl_error)?)
+    ))
+}
+
+fn openssl_error(err: ErrorStack) -> azure_core::error::Error {
+    Error::new(ErrorKind::Credential, err)
+}
+
+#[cfg_attr(target_arch = "wasm32", async_trait::async_trait(?Send))]
```
Comment (overvenus) [inline components/cloud/azure/src/token_credentials_with_cache/certificate_credentials.rs:164]: Why do we care about "wasm32"?

File: components/cloud/azure/src/token_credentials_with_cache/utils.rs [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+use azure_core::auth::TokenResponse;
+use time::OffsetDateTime;
+
+/// Refresh time to use in seconds
+pub(crate) const DEFAULT_REFRESH_TIME: i64 = 300;
+/// Timeout on refreshing token in seconds
+const DEFAULT_REFRESH_TOKEN_TIMEOUT: i64 = 10;
+
+fn config_default_refresh_token_timeout() -> i64 {
+    fail::fail_point!("mock_token_cache::refresh_timeout", |_| 1);
```
Comment (overvenus) [inline components/cloud/azure/src/token_credentials_with_cache/utils.rs:14]:
```suggestion
    fail::fail_point!("on_config_default_refresh_token_timeout", |_| 1);
```

In tikv, we prefer name failpoints in the style `[on_]func_name[_verb]`, "[ ]" means optional.

File: components/cloud/azure/src/token_credentials_with_cache/utils.rs [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+        let token_cache = TokenCache::default();
+        assert!(!token_cache.is_fresh());
+        let (token1, token2) = (token_cache.clone(), token_cache.clone());
+        let thread1 = thread::spawn(move || {
+            assert!(token1.need_update_token());
+            thread::sleep(Duration::from_secs(2));
+        });
+        let thread2 = thread::spawn(move || {
+            thread::sleep(Duration::from_millis(10));
+            let _f = fail::FailGuard::new("mock_token_cache::refresh_timeout", "return").unwrap();
```
Comment (overvenus) [inline components/cloud/azure/src/token_credentials_with_cache/utils.rs:109]: The tests requires `failpoints` is enabled.

File: components/cloud/azure/src/token_credentials_with_cache/certificate_credentials.rs [hunk 4]
```diff
@@ -0,0 +1,283 @@
...
+#[cfg_attr(not(target_arch = "wasm32"), async_trait::async_trait)]
+impl TokenCredential for ClientCertificateCredentialWithCache {
+    async fn get_token(&self, resource: &str) -> azure_core::Result<TokenResponse> {
```
Comment (overvenus) [inline components/cloud/azure/src/token_credentials_with_cache/certificate_credentials.rs:167]: Could you add docs and references about header, payload and other strings in the method?

Comment (LykxSassinator) [reply components/cloud/azure/src/token_credentials_with_cache/certificate_credentials.rs:167]: Yes, but i think what we've done for caching the `TokenCredential` is duplicated with the main branch, as the `Azure-sdk-for-rute` has already merged another committers' pr on this feature recently.

Comment (LykxSassinator) [reply components/cloud/azure/src/token_credentials_with_cache/certificate_credentials.rs:167]: I'll re-tidy this part.

File: components/cloud/azure/src/token_credentials_with_cache/certificate_credentials.rs (hunk 1 above)
Comment (LykxSassinator) [reply components/cloud/azure/src/token_credentials_with_cache/certificate_credentials.rs:164]: It's used to meet the requirement of the trait `TokenCredential`. As this feature has been supported in the latest version of `Azure-sdk-for-rust`, this part on caching `TokenCredential` can be skipped.

File: components/cloud/azure/src/token_credentials_with_cache/utils.rs (hunk 3 above)
Comment (LykxSassinator) [reply components/cloud/azure/src/token_credentials_with_cache/utils.rs]: Skipped. As this part on supporting caching `TokenCredential` is duplicated with the implementation on master branch.

File: components/cloud/azure/src/token_credentials_with_cache/utils.rs (hunk 2 above)
Comment (LykxSassinator) [reply components/cloud/azure/src/token_credentials_with_cache/utils.rs]: Skipped.

File: components/cloud/azure/src/token_credentials_with_cache/certificate_credentials.rs (hunk 4 above)
Comment (LykxSassinator) [reply components/cloud/azure/src/token_credentials_with_cache/certificate_credentials.rs:167]: It's just a formal released copy with caching feature from [`ClientCertificateCredentials`](https://github.com/Azure/azure-sdk-for-rust/blob/f90f7287dd4bfe4e32689cf1d2f8de3845344ca4/sdk/identity/src/token_credentials/client_certificate_credentials.rs#L149-L150), and it's proved to be safe by testing.

File: components/cloud/azure/src/token_credentials_with_cache/certificate_credentials.rs (hunk 1 above)
Comment (overvenus) [reply components/cloud/azure/src/token_credentials_with_cache/certificate_credentials.rs:164]: Could you add some comments?

File: components/cloud/azure/src/token_credentials_with_cache/secret_credentials.rs [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+use super::utils::{TokenCache, TokenCredentialWithCache};
+
+/// Enables authentication to Azure Active Directory using a client secret that
+/// was generated for an App Registration. It will automatically cache the
+/// latest token from Azure Active Directory.
+pub struct ClientSecretCredentialWithCache {
+    credential: ClientSecretCredential,
+
+    /// Cached token
+    cached_token: TokenCache,
```
Comment (overvenus) [inline components/cloud/azure/src/token_credentials_with_cache/secret_credentials.rs:20]: Why we need the cache?

**Review by overvenus — COMMENTED** [review]
Rest LGTM

File: components/cloud/azure/src/token_credentials_with_cache/secret_credentials.rs (hunk 5 above)
Comment (LykxSassinator) [reply components/cloud/azure/src/token_credentials_with_cache/secret_credentials.rs:20]: To reduce the cost of requesting a new TOKEN from Azure, this operation costs a lot (seconds level latency). 
And a new token returned from Azure keeps 1 hour ttl as default. That is, during the ttl, we could cache this token to reduce the cost on requesting new token.

File: components/cloud/azure/src/token_credentials_with_cache/certificate_credentials.rs (hunk 1 above)
Comment (LykxSassinator) [reply components/cloud/azure/src/token_credentials_with_cache/certificate_credentials.rs:166]: Done for it ,thx.

**Review by overvenus — APPROVED** [review]

File: components/cloud/azure/src/token_credentials_with_cache/utils.rs [hunk 6]
```diff
@@ -0,0 +1,121 @@
...
+    pub fn need_update_token(&self) -> bool {
+        let now = OffsetDateTime::now_utc().unix_timestamp();
+        loop {
+            if self.is_fresh() {
+                return false;
+            }
+            if let Ok(ref mut mtx) = self.wait_refreshed.lock() {
+                // Only allowed he first one who gets the lock to refresh the token.
+                if !**mtx {
+                    **mtx = true;
```
Comment (Leavrth) [inline components/cloud/azure/src/token_credentials_with_cache/utils.rs:55]: will here be `self.if_fresh()` is also true, according to function `update_token`?

**Review by Leavrth — APPROVED** [review]

**Review by overvenus — APPROVED** [review]

**Comment by LykxSassinator** [comment]
@Leavrth As `azure-sdk-for-rust` already supports TokenCredentials with cache by `AutoRefreshingTokenCredentials`, I've revoked the similar implementation on this part.

**Review by Leavrth — APPROVED** [review]

**Review by tonyxuqqi — APPROVED** [review]

**Pull request merged**